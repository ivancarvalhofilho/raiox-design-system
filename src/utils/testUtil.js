import { ThemeProvider } from 'styled-components'
import { Tokens } from '../tokens'
import React from 'react'
import { mount } from 'enzyme'
import 'jest-styled-components'

export function mountWithTheme(child) {
  return mount(child, {
    wrappingComponent: ({ children }) => (
      <ThemeProvider Tokens={Tokens}>{children}</ThemeProvider>
    ),
  })
}

export function camelize(str) {
  return str.replace(/(?:^\w|[A-Z]|\b\w)/g, function(word, index) {
    return index === 0 ? word.toLowerCase() : word.toUpperCase();
  }).replace(/\s+/g, '').replace(/\W/gm,'');
}
