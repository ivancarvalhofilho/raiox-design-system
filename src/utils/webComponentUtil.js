import * as React from 'react';
import PropTypes from "prop-types";
import ReactHtmlParser from "react-html-parser";
import {Tokens} from "../tokens";

// Converte os nomes de props do react para nomes válidos para o html
export const convertPropsToHtmlProps = (attributes) => {
    const aux = {}

    Object.keys(attributes).map(k => ({
        [k.replace(/(\B[A-Z])/gm, "-$1").toLowerCase()]: attributes[k]
    })).forEach(attribute => {
        aux[Object.keys(attribute)] = attribute[Object.keys(attribute)]
    })
    aux.child = PropTypes.element
    return aux
}

// Converte os nomes de props do react para nomes válidos para o html
export const mapHtmlPropsToReactProps = (originalProps, propValues) => {
    return Object.keys(originalProps).reduce((result,original) => {
        let htmlProp = original.replace(/(\B[A-Z])/gm, "-$1").toLowerCase()
        if(original === 'children')
           htmlProp = 'child'
        return {
            ...result,
            [original]: parseHtmlValue(propValues[htmlProp]),
        }
    },{})
}

// Converte strings em tipos do JS (int, bool, function, json, array)
const parseHtmlValue = (attrValue) => {
    if(attrValue == null) return attrValue
    if(typeof attrValue !== 'string') return attrValue

    let value = attrValue
    try {
        if (typeof eval(attrValue) === 'function') {
            value = () => eval(attrValue).call()
        }
        else throw 1
    } catch {
        if (!!attrValue.startsWith && attrValue.startsWith("icon:")) {
            attrValue = attrValue.replace("icon:","")
            value = Tokens.icons[attrValue]
        } else if (attrValue === 'true' || attrValue === 'false') {
            value = attrValue === 'true';
        } else if (!isNaN(attrValue) && attrValue !== '') {
            value = +attrValue;
        }
        else if (/^{.*}/.exec(attrValue) || /^\[[.\s\S]*]/.exec(attrValue)) {
            const badJson = attrValue
            const quotesOnKeys = badJson.replace(/(['"])?([a-z0-9A-Z_]+)(['"])?:/g, '"$2": ')
            const onlyDoubleQuotes = quotesOnKeys.replace(/'/gm, '\"')
            const removeTraillingComma = onlyDoubleQuotes.replace(/,[\s\\n]*\]/, "]")
            const goodJsonString = removeTraillingComma
            value = JSON.parse(goodJsonString);
        } else if(attrValue?.trim().startsWith("<") && !attrValue?.trim().startsWith("<svg")){
            value = ReactHtmlParser(attrValue)
        }
    }
    return value
}
