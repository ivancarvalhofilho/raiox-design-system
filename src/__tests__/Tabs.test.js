import React from 'react'
import toJson from 'enzyme-to-json'
import { mountWithTheme } from '../utils/testUtil'
import Tabs from '../components/Tabs/Tabs'

it('should render', () => {
  const spy = jest.spyOn(global.console, 'error')
  let props = {
    tabs: ['Categorias', 'Maquininhas', 'Banco'],
    tabActive: 0,
  }

  const app = mountWithTheme(
    <Tabs
      {...props}
      onTabClick={index => {
        props = { ...props, tabActive: index }
      }}
    />,
  )
  expect(toJson(app)).toMatchSnapshot()
  expect(spy).not.toHaveBeenCalled()
})

it('should render disabled', () => {
  const spy = jest.spyOn(global.console, 'error')
  let props = {
    tabs: ['Categorias', 'Maquininhas', 'Banco'],
    tabActive: 0,
    disabled: true,
  }

  const app = mountWithTheme(
    <Tabs
      {...props}
      onTabClick={index => {
        props = { ...props, tabActive: index }
      }}
    />,
  )
  expect(toJson(app)).toMatchSnapshot()
  expect(spy).not.toHaveBeenCalled()
})

it('should render with second tab', () => {
  const spy = jest.spyOn(global.console, 'error')
  let props = {
    tabs: ['Categorias', 'Maquininhas', 'Banco'],
    tabActive: 1,
  }

  const app = mountWithTheme(
    <Tabs
      {...props}
      onTabClick={index => {
        props = { ...props, tabActive: index }
      }}
    />,
  )
  expect(toJson(app)).toMatchSnapshot()
  expect(spy).not.toHaveBeenCalled()
})

it('should change tab', () => {
  const spy = jest.spyOn(global.console, 'error')
  const mockClick = jest.fn()
  let props = {
    tabs: ['Categorias', 'Maquininhas', 'Banco'],
    tabActive: 0,
  }

  const tabs = mountWithTheme(
    <Tabs
      {...props}
      onTabClick={index => {
        props = { ...props, tabActive: index }
        mockClick()
      }}
    />,
  )

  const tab1 = tabs.find('#tab1')
  tab1.first().simulate('click')
  tabs.setProps(props)
  expect(mockClick).toHaveBeenCalled()
  expect(tabs.props().tabActive).toBe(1)
  expect(spy).not.toHaveBeenCalled()
})

it('should not change tab', () => {
  const spy = jest.spyOn(global.console, 'error')
  const mockClick = jest.fn()
  let props = {
    tabs: ['Categorias', 'Maquininhas', 'Banco'],
    tabActive: 0,
    disabled: true,
  }

  const tabs = mountWithTheme(
    <Tabs
      {...props}
      onTabClick={index => {
        props = { ...props, tabActive: index }
        mockClick()
      }}
    />,
  )

  const tab1 = tabs.find('#tab1')
  tab1.first().simulate('click')
  tabs.setProps(props)
  expect(mockClick).not.toHaveBeenCalled()
  expect(tabs.props().tabActive).toBe(0)
  expect(spy).not.toHaveBeenCalled()
})
