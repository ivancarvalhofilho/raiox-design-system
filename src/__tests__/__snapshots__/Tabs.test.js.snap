// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
.c2 {
  padding: 5px 0px;
  font-family: Roboto;
  font-weight: 500;
  width: 33.333333333333336%;
  text-align: center;
  color: #008488;
  font-size: 12px;
}

.c3 {
  padding: 5px 0px;
  font-family: Roboto;
  font-weight: 400;
  width: 33.333333333333336%;
  text-align: center;
  color: #595A5C;
  font-size: 12px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
}

.c0 {
  cursor: pointer;
}

.c4 {
  width: 33.333333333333336%;
  height: 4px;
  border-radius: 4px 4px 0px 0px;
  background-color: #008488;
  -webkit-transform: translate(0%,0);
  -ms-transform: translate(0%,0);
  transform: translate(0%,0);
  -webkit-transition: -webkit-transform 0.3s linear;
  -webkit-transition: transform 0.3s linear;
  transition: transform 0.3s linear;
}

.c5 {
  width: 100%;
  height: 1px;
  background-color: #E0E0E0;
  -webkit-transform: translate(%,0);
  -ms-transform: translate(%,0);
  transform: translate(%,0);
  -webkit-transition: -webkit-transform 0.3s linear;
  -webkit-transition: transform 0.3s linear;
  transition: transform 0.3s linear;
}

<Tabs
  onTabClick={[Function]}
  tabActive={0}
  tabs={
    Array [
      "Categorias",
      "Maquininhas",
      "Banco",
    ]
  }
>
  <styled.div>
    <div
      className="c0"
    >
      <styled.div>
        <div
          className="c1"
        >
          <Tab
            active={true}
            id="tab0"
            key="0"
            label="Categorias"
            onClick={[Function]}
            width={33.333333333333336}
          >
            <styled.div
              active={true}
              id="tab0"
              onClick={[Function]}
              width={33.333333333333336}
            >
              <div
                className="c2"
                id="tab0"
                onClick={[Function]}
                width={33.333333333333336}
              >
                Categorias
              </div>
            </styled.div>
          </Tab>
          <Tab
            active={false}
            id="tab1"
            key="1"
            label="Maquininhas"
            onClick={[Function]}
            width={33.333333333333336}
          >
            <styled.div
              active={false}
              id="tab1"
              onClick={[Function]}
              width={33.333333333333336}
            >
              <div
                className="c3"
                id="tab1"
                onClick={[Function]}
                width={33.333333333333336}
              >
                Maquininhas
              </div>
            </styled.div>
          </Tab>
          <Tab
            active={false}
            id="tab2"
            key="2"
            label="Banco"
            onClick={[Function]}
            width={33.333333333333336}
          >
            <styled.div
              active={false}
              id="tab2"
              onClick={[Function]}
              width={33.333333333333336}
            >
              <div
                className="c3"
                id="tab2"
                onClick={[Function]}
                width={33.333333333333336}
              >
                Banco
              </div>
            </styled.div>
          </Tab>
        </div>
      </styled.div>
      <styled.div
        left={0}
        rounded={true}
        width={33.333333333333336}
      >
        <div
          className="c4"
          width={33.333333333333336}
        />
      </styled.div>
      <styled.div
        color="#E0E0E0"
        height={1}
        width={100}
      >
        <div
          className="c5"
          color="#E0E0E0"
          height={1}
          width={100}
        />
      </styled.div>
    </div>
  </styled.div>
</Tabs>
`;

exports[`should render disabled 1`] = `
.c2 {
  padding: 5px 0px;
  font-family: Roboto;
  font-weight: 500;
  width: 33.333333333333336%;
  text-align: center;
  color: #008488;
  font-size: 12px;
}

.c3 {
  padding: 5px 0px;
  font-family: Roboto;
  font-weight: 400;
  width: 33.333333333333336%;
  text-align: center;
  color: #595A5C;
  font-size: 12px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  pointer-events: none;
}

.c0 {
  opacity: 0.5;
  cursor: not-allowed;
}

.c4 {
  width: 33.333333333333336%;
  height: 4px;
  border-radius: 4px 4px 0px 0px;
  background-color: #008488;
  -webkit-transform: translate(0%,0);
  -ms-transform: translate(0%,0);
  transform: translate(0%,0);
  -webkit-transition: -webkit-transform 0.3s linear;
  -webkit-transition: transform 0.3s linear;
  transition: transform 0.3s linear;
}

.c5 {
  width: 100%;
  height: 1px;
  background-color: #E0E0E0;
  -webkit-transform: translate(%,0);
  -ms-transform: translate(%,0);
  transform: translate(%,0);
  -webkit-transition: -webkit-transform 0.3s linear;
  -webkit-transition: transform 0.3s linear;
  transition: transform 0.3s linear;
}

<Tabs
  disabled={true}
  onTabClick={[Function]}
  tabActive={0}
  tabs={
    Array [
      "Categorias",
      "Maquininhas",
      "Banco",
    ]
  }
>
  <styled.div
    disabled={true}
  >
    <div
      className="c0"
      disabled={true}
    >
      <styled.div
        disabled={true}
      >
        <div
          className="c1"
          disabled={true}
        >
          <Tab
            active={true}
            id="tab0"
            key="0"
            label="Categorias"
            onClick={[Function]}
            width={33.333333333333336}
          >
            <styled.div
              active={true}
              id="tab0"
              onClick={[Function]}
              width={33.333333333333336}
            >
              <div
                className="c2"
                id="tab0"
                onClick={[Function]}
                width={33.333333333333336}
              >
                Categorias
              </div>
            </styled.div>
          </Tab>
          <Tab
            active={false}
            id="tab1"
            key="1"
            label="Maquininhas"
            onClick={[Function]}
            width={33.333333333333336}
          >
            <styled.div
              active={false}
              id="tab1"
              onClick={[Function]}
              width={33.333333333333336}
            >
              <div
                className="c3"
                id="tab1"
                onClick={[Function]}
                width={33.333333333333336}
              >
                Maquininhas
              </div>
            </styled.div>
          </Tab>
          <Tab
            active={false}
            id="tab2"
            key="2"
            label="Banco"
            onClick={[Function]}
            width={33.333333333333336}
          >
            <styled.div
              active={false}
              id="tab2"
              onClick={[Function]}
              width={33.333333333333336}
            >
              <div
                className="c3"
                id="tab2"
                onClick={[Function]}
                width={33.333333333333336}
              >
                Banco
              </div>
            </styled.div>
          </Tab>
        </div>
      </styled.div>
      <styled.div
        left={0}
        rounded={true}
        width={33.333333333333336}
      >
        <div
          className="c4"
          width={33.333333333333336}
        />
      </styled.div>
      <styled.div
        color="#E0E0E0"
        height={1}
        width={100}
      >
        <div
          className="c5"
          color="#E0E0E0"
          height={1}
          width={100}
        />
      </styled.div>
    </div>
  </styled.div>
</Tabs>
`;

exports[`should render with second tab 1`] = `
.c2 {
  padding: 5px 0px;
  font-family: Roboto;
  font-weight: 400;
  width: 33.333333333333336%;
  text-align: center;
  color: #595A5C;
  font-size: 12px;
}

.c3 {
  padding: 5px 0px;
  font-family: Roboto;
  font-weight: 500;
  width: 33.333333333333336%;
  text-align: center;
  color: #008488;
  font-size: 12px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
}

.c0 {
  cursor: pointer;
}

.c4 {
  width: 33.333333333333336%;
  height: 4px;
  border-radius: 4px 4px 0px 0px;
  background-color: #008488;
  -webkit-transform: translate(100%,0);
  -ms-transform: translate(100%,0);
  transform: translate(100%,0);
  -webkit-transition: -webkit-transform 0.3s linear;
  -webkit-transition: transform 0.3s linear;
  transition: transform 0.3s linear;
}

.c5 {
  width: 100%;
  height: 1px;
  background-color: #E0E0E0;
  -webkit-transform: translate(%,0);
  -ms-transform: translate(%,0);
  transform: translate(%,0);
  -webkit-transition: -webkit-transform 0.3s linear;
  -webkit-transition: transform 0.3s linear;
  transition: transform 0.3s linear;
}

<Tabs
  onTabClick={[Function]}
  tabActive={1}
  tabs={
    Array [
      "Categorias",
      "Maquininhas",
      "Banco",
    ]
  }
>
  <styled.div>
    <div
      className="c0"
    >
      <styled.div>
        <div
          className="c1"
        >
          <Tab
            active={false}
            id="tab0"
            key="0"
            label="Categorias"
            onClick={[Function]}
            width={33.333333333333336}
          >
            <styled.div
              active={false}
              id="tab0"
              onClick={[Function]}
              width={33.333333333333336}
            >
              <div
                className="c2"
                id="tab0"
                onClick={[Function]}
                width={33.333333333333336}
              >
                Categorias
              </div>
            </styled.div>
          </Tab>
          <Tab
            active={true}
            id="tab1"
            key="1"
            label="Maquininhas"
            onClick={[Function]}
            width={33.333333333333336}
          >
            <styled.div
              active={true}
              id="tab1"
              onClick={[Function]}
              width={33.333333333333336}
            >
              <div
                className="c3"
                id="tab1"
                onClick={[Function]}
                width={33.333333333333336}
              >
                Maquininhas
              </div>
            </styled.div>
          </Tab>
          <Tab
            active={false}
            id="tab2"
            key="2"
            label="Banco"
            onClick={[Function]}
            width={33.333333333333336}
          >
            <styled.div
              active={false}
              id="tab2"
              onClick={[Function]}
              width={33.333333333333336}
            >
              <div
                className="c2"
                id="tab2"
                onClick={[Function]}
                width={33.333333333333336}
              >
                Banco
              </div>
            </styled.div>
          </Tab>
        </div>
      </styled.div>
      <styled.div
        left={100}
        rounded={true}
        width={33.333333333333336}
      >
        <div
          className="c4"
          width={33.333333333333336}
        />
      </styled.div>
      <styled.div
        color="#E0E0E0"
        height={1}
        width={100}
      >
        <div
          className="c5"
          color="#E0E0E0"
          height={1}
          width={100}
        />
      </styled.div>
    </div>
  </styled.div>
</Tabs>
`;
